shard/committee/assignment.go

```diff
@@ -116,62 +125,148 @@ func preStakingEnabledCommittee(s shardingconfig.Instance) shard.State {
-			superComm[spot].NodeList,
-			shard.NodeID{
-				top[i].Address,
-				pubKey,
-				&shard.StakedMember{big.NewInt(0)},
-			},
-		)
+		pubKey.FromLibBLSPublicKey(pub)
+		superComm[spot].NodeList[fillCount[spot]] = shard.NodeID{
+			common2.ParseAddr(hAccounts[i].Address),
+			pubKey,
+			nil,
+		}
+		fillCount[spot]++
+	}
+
+	staked := effective.Apply(essentials)
+
+	sort.SliceStable(
+		staked,
+		func(i, j int) bool { return staked[i].Dec.GTE(staked[j].Dec) },
+	)
+
+	shardBig := big.NewInt(int64(shardCount))
+
+	if len(staked) <= stakedSlotsCount {
+		// WARN unlikely to happen in production but will happen as we are developing
+	}
+
+	for i := 0; i < stakedSlotsCount; i++ {
+		bucket := int(new(big.Int).Mod(staked[i].Address.Big(), shardBig).Int64())
+		slot := staked[i]
+		pubKey := essentials[slot.Address].SpreadAmong[slotUsage[slot.Address]-1]
+		slotUsage[slot.Address]--
+		// Keep going round till find an open spot
+		for j := bucket; ; j = (j + 1) % shardCount {
+			if fillCount[j] != maxNodePerShard {
+				superComm[j].NodeList[fillCount[j]] = shard.NodeID{
+					slot.Address,
+					pubKey,
+					&slot.Dec,
+				}
+				fillCount[j]++
+				break
+			}
+		}
 	}
 
-	utils.Logger().Info().Ints("distribution of Stakers in Shards", fillCount)
 	return superComm, nil
 }
 
+// ComputePublicKeys produces publicKeys of entire supercommittee per epoch, optionally providing a
+// shard specific subcommittee
+func (def partialStakingEnabled) ComputePublicKeys(
+	epoch *big.Int, d DataProvider, shardID int,
+) ([]*bls.PublicKey, []*bls.PublicKey) {
+	config := d.Config()
+	instance := shard.Schedule.InstanceForEpoch(epoch)
+	superComm := shard.State{}
+	stakedSlots :=
+		(instance.NumNodesPerShard() - instance.NumHarmonyOperatedNodesPerShard()) *
+			int(instance.NumShards())
+
+	if config.IsStaking(epoch) {
+		superComm, _ = eposStakedCommittee(instance, d, stakedSlots)
+	} else {
+		superComm = preStakingEnabledCommittee(instance)
+	}
+
+	spot := 0
+	shouldBe := int(instance.NumShards()) * instance.NumNodesPerShard()
+
+	total := 0
+	for i := range superComm {
+		total += len(superComm[i].NodeList)
+	}
+
+	if shouldBe != total {
+		fmt.Println("Count mismatch", shouldBe, total)
+	}
+
+	allIdentities := make([]*bls.PublicKey, shouldBe)
+	for i := range superComm {
+		for j := range superComm[i].NodeList {
+			identity := &bls.PublicKey{}
+			superComm[i].NodeList[j].BlsPublicKey.ToLibBLSPublicKey(identity)
+			allIdentities[spot] = identity
+			spot++
+		}
+	}
+
+	if shardID == StateID {
+		return allIdentities, nil
+	}
+
+	subCommittee := superComm.FindCommitteeByID(uint32(shardID))
+	subCommitteeIdentities := make([]*bls.PublicKey, len(subCommittee.NodeList))
+	spot = 0
+	for i := range subCommittee.NodeList {
+		identity := &bls.PublicKey{}
+		subCommittee.NodeList[i].BlsPublicKey.ToLibBLSPublicKey(identity)
+		subCommitteeIdentities[spot] = identity
+		spot++
+	}
+
+	return allIdentities, subCommitteeIdentities
```

Comment:
Why do we need allIdentities as a list? Why just return all identities grouped by shard id and remove the subCommitteeIdentities?

Reply:
because caller has to then find their shardID rather than ignore the data they don't need

Reply:
Remove the flattened version and group by shards.